lib/utils/fs.go

```diff
@@ -241,32 +242,82 @@ func FSTryReadLockTimeout(ctx context.Context, filePath string, timeout time.Dur
-func RemoveSecure(filePath string) error {
-	for i := 0; i < 3; i++ {
-		if err := overwriteFile(filePath); err != nil {
-			return trace.Wrap(err)
+// RemoveAllSecure is similar to [os.RemoveAll] but leverages [RemoveSecure] to delete files.
+func RemoveAllSecure(path string) error {
+	// Match os.RemoveAll protections in not permitting removal of "." directories
+	// This check comes directly from https://cs.opensource.google/go/go/+/refs/tags/go1.21.1:src/os/removeall_at.go;l=24
+	if path == "." || (len(path) >= 2 && path[len(path)-1] == '.' && os.IsPathSeparator(path[len(path)-2])) {
+		return &os.PathError{Op: "RemoveAllSecure", Path: path, Err: syscall.EINVAL} // error type matches os.RemoveAll
+	}
+
+	info, err := os.Lstat(path)
+	switch {
+	case err != nil && os.IsNotExist(err):
+		return nil
+	case err != nil:
+		return trace.ConvertSystemError(err)
+	case !info.IsDir():
+		return RemoveSecure(path)
+	}
+	var removeErrors []error
+	files, err := os.ReadDir(path)
+	if err != nil {
+		// Don't fail fast, allow removal at end to be attempted.
+		removeErrors = []error{err}
+	}
+	// It's possible for a partial file list to be returned even if an error above was returned.
+	for _, f := range files {
+		if err := RemoveAllSecure(filepath.Join(path, f.Name())); err != nil {
+			removeErrors = append(removeErrors, err)
 		}
 	}
-	return trace.ConvertSystemError(os.Remove(filePath))
+	if err := os.Remove(path); err != nil {
+		removeErrors = append(removeErrors, err)
+	}
+	switch len(removeErrors) {
+	case 1:
+		return trace.ConvertSystemError(removeErrors[0])
+	case 0:
+		return nil
+	default:
+		return trace.NewAggregate(removeErrors...)
+	}
 }
 
-func overwriteFile(filePath string) (err error) {
-	f, err := os.OpenFile(filePath, os.O_WRONLY, 0)
-	if err != nil {
-		return trace.ConvertSystemError(err)
+// RemoveSecure attempts to securely delete the file by first overwriting the file with random data three times
+// followed by calling os.Remove(filePath).
+func RemoveSecure(filePath string) error {
+	f, openErr := os.OpenFile(filePath, os.O_WRONLY, 0)
+	if os.IsNotExist(openErr) {
+		return openErr
+	} else if openErr == nil {
+		defer func() {
+			if closeErr := f.Close(); closeErr != nil {
+				log.WithError(closeErr).Warningf("Failed to close deleted file %v.", f.Name())
+			}
+		}()
 	}
-	defer func() {
-		if closeErr := f.Close(); closeErr != nil {
-			if err == nil {
-				err = trace.ConvertSystemError(closeErr)
-			} else {
-				log.WithError(closeErr).Warningf("Failed to close %v.", f.Name())
+
+	if runtime.GOOS == "windows" {
+		// windows can't unlink the file before overwriting
+		if f != nil {
+			for i := 0; i < 3; i++ {
+				_ = overwriteFile(f)
 			}
 		}
-	}()
+		return trace.ConvertSystemError(os.Remove(filePath))
+	} else {
+		removeErr := os.Remove(filePath)
+		if f != nil {
+			for i := 0; i < 3; i++ {
+				_ = overwriteFile(f)
+			}
+		}
+		return trace.ConvertSystemError(removeErr)
+	}
+}
```

Comment:
Suggestion: drop separate branches. Builds on the previous suggestions by assuming `f` can't be nil in this block.

```suggestion
	isWindows := runtime.GOOS == "windows"
	var removeErr error
	if !isWindows {
		// Unlink first on unix systems.
		removeErr = os.Remove(filePath)
	}
	
	for i := 0; i < 3; i++ {
		_ = overwriteFile(f)
	}

	if isWindows {
		removeErr = os.Remove(filePath)
	}
		
	return trace.ConvertSystemError(removeErr)
}
```

Reply:
Mixing the conditional multiple times reduces readability to me, but I don't feel strongly about it

Reply:
Both look fine to me.